File: hashtable/hashtable.js
// implement hash table by ES6 class syntax
class Node {
	constructor(value, next = null) {
		this.value = value;
		this.next = next;
	}
}

class MyLinkedList {
	constructor(head = null) {
		this.head = head;
	}
}

class Data {
	constructor(value, flag, key) {
		this.value = value;
		this.flag = flag;
		this.key = key;
	}
}

class MyHashTable {
	constructor(size) {
		this.hashtable = new Array(size);
		for (let i = 0; i < this.hashtable.length; i += 1) {
			this.hashtable[i] = null;
		}
	}	

	// hash function
	hash(value) {
		const key = value % this.hashtable.length;
		return key;
	}

	// double hashing
	doubleHash(value, primeNumber) {
		const firstKey = this.hash(value);
		const secondKey = primeNumber - (value % primeNumber);
		const key = firstKey + secondKey;
		return key;
	}

	// insert a element by chaining
	chainingInsert(value) {
		const key = this.hash(value);
		if (this.hashtable[key] === null) {
			const node = new Node(value);
			let linkedlist = new MyLinkedList();
			linkedlist.head = node;
			this.hashtable[key] = linkedlist;
		} else {
			const node = new Node(value);
			let p = this.hashtable[key].head;
			while (p.next !== null) {
				p = p.next;
			}
			p.next = node;
		}
	}

	// searching a element on chains
	chainingSearch(value) {
		const key = this.hash(value);
		if (this.hashtable[key] === null) return false;
		else {
			let p = this.hashtable[key].head;
			while (p !== null) {
				if (p.value === value) return true;
				else p = p.next;
			}
			return false;
		}
	}

	// delete a element on chains
	chainingDelete(value) {
		const key = this.hash(value);
		if (this.hashtable[key] === null) return;
		else {
			let p = this.hashtable[key].head;
			if (p.value === value) {
				this.hashtable[key].head = this.hashtable[key].head.next;
				if (this.hashtable[key].head === null) this.hashtable[key] = null;
				return;
			} else {
				while (p.next !== null && p !== null) {
					if (p.next.value === value) {
						p.next = p.next.next;
						return;
					} else p = p.next;
				}
			}
		}
	}

	// insert a element by linear probing
	linearProbingInsert(value) {
		const key = this.hash(value);
		if (this.hashtable[key] === null) {
			const data = new Data(value, true, key);
			this.hashtable[key] = data;
		} else {
			for (let i = key + 1; i < this.hashtable.length; i += 1) {
				if (this.hashtable[i] === null) {
					const data = new Data(value, false, key);
					this.hashtable[i] = data;
					return true;
				}
			}
			for (let i = 0; i < key; i += 1) {
				if (this.hashtable[i] === null) {
					const data = new Data(value, false, key);
					this.hashtable[i] = data;
					return true;
				}	
			}
			return false;
		}
	}

	// searching a element by linear probing
	linearProbingSearch(value) {
		const key = this.hash(value);
		if (this.hashtable[key] === null) return false;
		else {
			if (this.hashtable[key].value === value) return true;
			else {
				for (let i = key + 1; i < this.hashtable.length; i += 1) {
					if (this.hashtable[i] === null) continue;
					else if (this.hashtable[i].value === value) return true;
				}
				for (let i = 0; i < key; i += 1) {
					if (this.hashtable[i] === null) continue;
					else if (this.hashtable[i].value === value) return true;
				}
				return false;	
			}
		}
	}

	// delete a element by linear probing
	linearProbingDelete(value) {
		const key = this.hash(value);
		if (this.hashtable[key] === null) return false;
		else {
			if (this.hashtable[key].value === value) {
				this.hashtable[key] = null;
				return true;
			} else {
				for (let i = key + 1; i < this.hashtable.length; i += 1) {
					if (this.hashtable[i] === null) continue;
					else if (this.hashtable[i].value === value) {
						this.hashtable[i] = null;
						return true;
					}
				}
				for (let i = 0; i < key; i += 1) {
					if (this.hashtable[i] === null) continue;
					else if (this.hashtable[i].value === value) {
						this.hashtable[i] = null;
						return true;
					}
				}
				return false;	
			}
		}	
	}

	// insert a element by quadratic probing
	QuadraticProbingInsert(value) {
		const key = this.hash(value);
		if (this.hashtable[key] === null) {
			const data = new Data(value, true, key);
			this.hashtable[key] = data;
		} else {
			for (let i = key + 1; i < this.hashtable.length; i += 2) {
				if (this.hashtable[i] === null) {
					const data = new Data(value, false, key);
					this.hashtable[i] = data;
					return true;
				}
			}
			for (let i = 0; i < key; i += 2) {
				if (this.hashtable[i] === null) {
					const data = new Data(value, false, key);
					this.hashtable[i] = data;
					return true;
				}	
			}
			return false;
		}
	}

	// searching a element by quadratic probing
	QuadraticProbingSearch(value) {
		const key = this.hash(value);
		if (this.hashtable[key] === null) return false;
		else {
			if (this.hashtable[key].value === value) return true;
			else {
				for (let i = key + 1; i < this.hashtable.length; i += 2) {
					if (this.hashtable[i] === null) continue;
					else if (this.hashtable[i].value === value) return true;
				}
				for (let i = 0; i < key; i += 2) {
					if (this.hashtable[i] === null) continue;
					else if (this.hashtable[i].value === value) return true;
				}
				return false;	
			}
		}
	}

	// delete a element by quadratic probing
	QuadraticProbingDelete(value) {
		const key = this.hash(value);
		if (this.hashtable[key] === null) return false;
		else {
			if (this.hashtable[key].value === value) {
				this.hashtable[key] = null;
				return true;
			} else {
				for (let i = key + 1; i < this.hashtable.length; i += 2) {
					if (this.hashtable[i] === null) continue;
					else if (this.hashtable[i].value === value) {
						this.hashtable[i] = null;
						return true;
					}
				}
				for (let i = 0; i < key; i += 2) {
					if (this.hashtable[i] === null) continue;
					else if (this.hashtable[i].value === value) {
						this.hashtable[i] = null;
						return true;
					}
				}
				return false;	
			}
		}	
	}
}

File: hashtable/hashtable.py
from typing import Any

class Node:
	def __init__(self, value: Any, next: 'Node'= None) -> None:
		self.value: Any = value 
		self.next: 'Node' = next 

class MyLinkedList:
	def __init__(self, head: 'Node'= None) -> None:
		self.head: 'Node' = head 

class Data:
	def __init__(self, value: Any, flag: bool, key: Any) -> None:
		self.value: Any = value 
		self.flag: bool = flag
		self.key: Any = key

class MyHashTable:
	def __init__(self, size: int) -> None:
		self.hashtable = list()
		for i in range(size):
			self.hashtable.append(None)

	def hash(self, value: Any) -> Any:
		'''
		hash function
		'''
		key = value % len(self.hashtable)
		return key

	def double_hash(self, value: Any, prime_number: int) -> Any:
		'''
		double hashing
		'''
		first_key = self.hash(value)
		second_key = prime_number - (value % prime_number)
		key = first_key + second_key
		return key

	def chaining_insert(self, value: Any) -> None:
		'''
		insert a element by chaining
		'''
		key = self.hash(value)
		if self.hashtable[key] == None:
			node = Node(value)
			linked_list = MyLinkedList()
			linked_list.head = node 
			self.hashtable[key] = linked_list
		else:
			node = Node(value)
			p = self.hashtable[key].head
			while p.next != None:
				p = p.next 
			p.next = node

	def chaining_search(self, value: Any) -> bool:
		'''
		searching a element on chains
		'''
		key = self.hash(value)
		if self.hashtable[key] == None:
			return False
		else:
			p = self.hashtable[key].head
			while p != None:
				if p.value == value:
					return True
				else:
					p = p.next
			return False

	def chaining_delete(self, value: Any) -> None:
		'''
		delete a element on chains
		'''
		key = self.hash(value)
		if self.hashtable[key] == None:
			return
		else:
			p = self.hashtable[key].head
			if p.value == value:
				self.hashtable[key].head = self.hashtable[key].head.next 
				if self.hashtable[key].head == None:
					self.hashtable[key] = None
					return
			else:
				while p.next != None and p != None:
					if p.next.value == value:
						p.next = p.next.next 
						return
					else:
						p = p.next 

	def linear_probling_insert(self, value: Any) -> bool:
		'''
		insert a element by linear probing
		'''
		key = self.hash(value)
		if self.hashtable[key] == None:
			data = Data(value, True, key)
			self.hashtable[key] = data
		else:
			for i in range(key + 1, len(self.hashtable)):
				if self.hashtable[i] == None:
					data = Data(value, False, key)
					self.hashtable[i] = data
					return True
			for i in range(key):
				if self.hashtable[i] == None:
					data = Data(value, False, key)
					self.hashtable[i] = data
					return True
			return False

	def linear_probling_search(self, value: Any) -> bool:
		'''
		searching a element by linear probing
		'''
		key = self.hash(value)
		if self.hashtable[key] == None:
			return False
		else:
			if self.hashtable[key].value == value:
				return True
			else:
				for i in range(key + 1, len(self.hashtable)):
					if self.hashtable[i] == None:
						continue
					elif self.hashtable[i].value == value:
						return True
				for i in range(key):
					if self.hashtable[i] == None:
						continue
					elif self.hashtable[i].value == value:
						return True
				return False

	def linear_probling_delete(self, value: Any) -> bool:
		'''
		delete a element by linear probing
		'''
		key = self.hash(value)
		if self.hashtable[key] == None:
			return False
		else:
			if self.hashtable[key].value == value:
				self.hashtable[key] = None
				return True
			else:
				for i in range(key + 1, len(self.hashtable)):
					if self.hashtable[i] == None:
						continue
					elif self.hashtable[i].value == value:
						self.hashtable[i] = None
						return True
				for i in range(key):
					if self.hashtable[i] == None:
						continue
					elif self.hashtable[i].value == value:
						self.hashtable[i] = None
						return True
				return False

	def quadratic_probling_insert(self, value: Any) -> bool:
		'''
		insert a element by quadratic probing
		'''
		key = self.hash(value)
		if self.hashtable[key] == None:
			data = Data(value, True, key)
			self.hashtable[key] = data
		else:
			for i in range(key + 1, len(self.hashtable), 2):
				if self.hashtable[i] == None:
					data = Data(value, False, key)
					self.hashtable[i] = data
					return True
			for i in range(0, key, 2):
				if self.hashtable[i] == None:
					data = Data(value, False, key)
					self.hashtable[i] = data
					return True
			return False

	def quadratic_probling_search(self, value: Any) -> bool:
		'''
		searching a element by quadratic probing
		'''
		key = self.hash(value)
		if self.hashtable[key] == None:
			return False
		else:
			if self.hashtable[key].value == value:
				return True
			else:
				for i in range(key + 1, len(self.hashtable), 2):
					if self.hashtable[i] == None:
						continue
					elif self.hashtable[i].value == value:
						return True
				for i in range(0, key, 2):
					if self.hashtable[i] == None:
						continue
					elif self.hashtable[i].value == value:
						return True
				return False

	def quadratic_probling_delete(self, value: Any) -> bool:
		'''
		delete a element by quadratic probing
		'''
		key = self.hash(value)
		if self.hashtable[key] == None:
			return False
		else:
			if self.hashtable[key].value == value:
				self.hashtable[key] = None
				return True
			else:
				for i in range(key + 1, len(self.hashtable), 2):
					if self.hashtable[i] == None:
						continue
					elif self.hashtable[i].value == value:
						self.hashtable[i] = None
						return True
				for i in range(0, key, 2):
					if self.hashtable[i] == None:
						continue
					elif self.hashtable[i].value == value:
						self.hashtable[i] = None
						return True
				return False